### ISSUE TITLE
System has not been booted with systemd as init system (PID 1). Can't operate. Failed to connect to bus: Host is down

### ISSUE THREAD

**Issue by TylerJaacks** [microsoft/WSL#8883]
### Version

Microsoft Windows [Version 10.0.22621.521]

### WSL Version

- [X] WSL 2
- [ ] WSL 1

### Kernel Version

5.15.57.1

### Distro Version

Ubuntu 22.04.1 LTS

### Other Software

N/A

### Repro Steps

Enable ```systemd``` via ```/etc/wsl.conf```

Run ```sudo systemctl list-units --type=service```

### Expected Behavior

List of services show.

### Actual Behavior

System has not been booted with systemd as init system (PID 1). Can't operate.
Failed to connect to bus: Host is down

**Comment by TylerJaacks** [comment]
```┌─(~)────────────────────────────────────────────────────────────────────────────────(tylerjaacks@TYLERJ-WIN11:pts/0)─┐
└─(18:13:37)──> cat /etc/wsl.conf                                                                       ──(Sat,Sep24)─┘
[boot]
systemd=true```

**Comment by elsaco** [comment]
@TylerJaacks based on the kernel version you posted, you're using `WSL 0.65.1`. In order to use `systemd` you'll have to upgrade to latest release, `WSL 0.67.6`. What is the output of `wsl --version` or `wsl --status`?

**Comment by TylerJaacks** [comment]
Yeah so I just figured it out now whenever I run the command and it just takes a while and gives me ```Failed to list units: Transport endpoint is not connected```

**Comment by tusharsnn** [comment]
> @TylerJaacks based on the kernel version you posted, you're using `WSL 0.65.1`. In order to use `systemd` you'll have to upgrade to latest release, `WSL 0.67.6`. What is the output of `wsl --version` or `wsl --status`?

If I am not wrong then it's only available to insiders for now, or is it available for all?

**Comment by echoechoin** [comment]
download 0.67.6 in `https://github.com/microsoft/WSL/releases` page

**Comment by TylerJaacks** [comment]
So I just gave up on this Preview and will wait for Stable cause it works fine on Stable Windows 11 Branch and then Systemd just breaks and If I try to run ```wsl.exe --shutdown``` I get the hanging computer issue.

**Comment by ejazahm3d** [comment]
So I have the same issue. It was working previously so I have no idea why it stopped working. I am using version from github releases. Here is the output for --version. 

```WSL version: 0.68.2.0
Kernel version: 5.15.62.1
WSLg version: 1.0.44
MSRDC version: 1.2.3401
Direct3D version: 1.606.4
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.22000.1042```

**Comment by scruel** [comment]
Same for wsl2 with Ubuntu 22.04.1 LTS
For `wsl --version`, I can only get help info, rather than version info... why?
For `wsl --status`, got `Kernel version: 5.10.102.1`, last updated at 2022/11/21.

**Comment by scruel** [comment]
Problem solved:
Goto https://github.com/microsoft/WSL/releases, download new version of WSL (0.67.6 and higher), install and reboot your system.
Enable systemd via `/etc/wsl.conf`
Then retry.

Thanks to @echoechoin.

**Comment by yilmazdurmaz** [comment]
Although the question was asked for Windows 11, the same error also comes up in Windows 10 installations if some updates are missing. These steps are for W10 users.
- Install WSL from within Windows Store
- Install update KB5020030. (I had to download it manually from the [Update Catalog](https://www.catalog.update.microsoft.com/Search.aspx?q=KB5020030))

This "install from Store" might also solve W11 problems. Check that first if this error happens.

**Comment by luc-languagetools** [comment]
I'm having the same issue on windows 10 pro
```
C:\WINDOWS\system32>wsl --version
WSL version: 1.0.0.0
Kernel version: 5.15.74.2
WSLg version: 1.0.47
MSRDC version: 1.2.3575
Direct3D version: 1.606.4
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.19045.2311
```

```
luc@wsl2-ubuntu$ systemctl status cron
System has not been booted with systemd as init system (PID 1). Can't operate.
Failed to connect to bus: Host is down
```

**Comment by yilmazdurmaz** [comment]
@luc-languagetools your system seems ready to use it. I am guessing you have forgotten to activate it in `/etc/wsl.conf`. please read answers in this thread to see how you do it.

**Comment by luc-languagetools** [comment]
I already have
```
[boot]
systemd=true
```
in `.wslconfig`

**Comment by yilmazdurmaz** [comment]
it is not `%userprofile%\.wslconfig` you would be editing. it is `/etc/wsl.config` inside. you will need to have root access
````
wsl  # in cmd or powershell
sudo nano /etc/wsl.conf # or use your fav editor
  #insert those 2 lines you have
exit
wsl --shutdown #  first shutdown
wsl # then restart
````
PS: this `.wslconfig` thing might be something that wsl team have in mind. at least for now, that is the other file you need to edit.

**Comment by luc-languagetools** [comment]
@yilmazdurmaz you are correct, I had been mistakenly editing %USERPROFILE%\.wslconfig
```
[boot]
systemd=true
```
should go into `/etc/wsl.conf` on the filesystem inside WSL. And now it works ! I really appreciate your help !

**Comment by Oluadepe** [comment]
@yilmazdurmaz  @luc-languagetools I cant find the /etc/wsl.conf on the filesystem, do I need to create it myself

**Comment by elsaco** [comment]
@Oluadepe yes, if `/etc/wsl.conf` is not present already create a new file and add `systemd` setting to it.

**Comment by Oluadepe** [comment]
> @Oluadepe yes, if `/etc/wsl.conf` is not present already create a new file and add `systemd` setting to it.

I just did and it worked. 🎉🎉🎉

**Comment by TylerJaacks** [comment]
Is there a solution for this because there appears to be two different posts here?

**Comment by yilmazdurmaz** [comment]
Let me unify the answers
- open a terminal and run this command: `wsl --version`. If you see "WSL version: 1.0.0.0" (or higher in 1.x.x.x) then you are ready to go. Else, follow the installation instructions until you see it. (Installing KB5020030 might be needed as discussed in https://github.com/microsoft/WSL/issues/9157)
- now log into WSL by `wsl` command in the same or new windows console (or `ubuntu` for example, if you have many distributions)
- inside the wsl shell, edit `/etc/wsl.conf` file as root or with sudo: `sudo nano /etc/wsl.conf`` (use your favorite linux editor)
- insert these 2 lines
````
[boot]
systemd=true
````
- `exit` the shell
- shutdown wsl: `wsl --shutdown`
- then restart it again: `wsl`.

**Comment by Luxpy** [comment]
Распределение по умолчанию: Ubuntu-20.04
Версия по умолчанию: 2 (WSL-2)
Версия ядра: 5.10.102.1 


I got an error:

```
sudo systemctl status postgresql

System has not been booted with systemd as init system (PID 1). Can't operate.
Failed to connect to bus: Host is down
```

I did all the things above:
```
cat /etc/wsl.conf
[boot]
systemd=true
```

I did run ` wsl --shutdown` to restart

**Comment by elsaco** [comment]
@Luxpy what version of WSL are you running? `systemd` support was added in https://github.com/microsoft/WSL/releases/tag/0.67.6, the same release that introduced Microsoft.WSL.Kernel version 5.15.62.1. Looks like you're running an older kernel implying an older WSL.

I can confirm that systemd works okay in Ubuntu 20.04 and WSL 1.0.0:

```
elsaco@RIPPER:~$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.3 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.3 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
elsaco@RIPPER:~$ systemctl status --no-pager
● RIPPER
    State: degraded
     Jobs: 0 queued
   Failed: 6 units
    Since: Mon 2022-12-12 08:22:02 PST; 19s ago
   CGroup: /
           ├─user.slice
           │ └─user-1000.slice
           │   ├─user@1000.service
           │   │ ├─session.slice
           │   │ │ ├─pipewire-media-session.service
           │   │ │ └─pipewire.service
           │   │ ├─app.slice
           │   │ │ ├─at-spi-dbus-bus.service
           │   │ │ └─dbus.service
           │   │ └─init.scope
           │   └─session-c1.scope
           ├─init.scope
           └─system.slice
---cut---
```

**Comment by yilmazdurmaz** [comment]
check this link if you haven't yet. https://devblogs.microsoft.com/commandline/systemd-support-is-now-available-in-wsl/
> **Ensuring you are on the right WSL version**
> This change is only available in the "Microsoft Store version of WSL" version 0.67.6 and higher. You can check your version number by running wsl --version. If that command fails then you are running the in-Windows version of WSL and need to [upgrade to the Store version](https://devblogs.microsoft.com/commandline/a-preview-of-wsl-in-the-microsoft-store-is-now-available/#how-to-install-and-use-wsl-in-the-microsoft-store)

You need to have the "Store version" from now on. Open "Microsoft Store" app and search "Windows Subsystem for Linux". I forgot where I read but there was a statement like "WSL will continue with the store version and receive updates from there". and its latest version has the following signature (at least mine shows no update available). Also use `wsl --update` to make sure of getting fixes. my current version shows these lines.
````
WSL version: 1.0.3.0
Kernel version: 5.15.79.1
````

for a few other important details, heck my above post: https://github.com/microsoft/WSL/issues/8883#issuecomment-1345472665

**Comment by Tamoghna-Saha** [comment]
> Let me unify the answers
> 
>     * open a terminal and run this command: `wsl --version`. If you see "WSL version: 1.0.0.0" (or higher in 1.x.x.x) then you are ready to go. Else, follow the installation instructions until you see it. (Installing KB5020030 might be needed as discussed in [WSL v1.0.0 cannot be installed on Windows 10 Version 19045, whereas supported on the System Requirements on Microsoft Store #9157](https://github.com/microsoft/WSL/issues/9157))
> 
>     * now log into WSL by `wsl` command in the same or new windows console (or `ubuntu` for example, if you have many distributions)
> 
>     * inside the wsl shell, edit `/etc/wsl.conf` file as root or with sudo: `sudo nano /etc/wsl.conf` (use your favorite linux editor)
> 
>     * insert these 2 lines
> 
> 
> ```
> [boot]
> systemd=true
> ```
> 
>     * `exit` the shell
> 
>     * shutdown wsl: `wsl --shutdown`
> 
>     * then restart it again: `wsl`.

Yup. This worked like a charm! Thanks

**Comment by mert-kurttutan** [comment]
> wsl --version

Hi Yilmaz,
What if we don't have access to store version (as it is blocked by company's IT policy in my case).
Do you have any suggestion for this problem?

Thanks in advance.

**Comment by yilmazdurmaz** [comment]
> What if we don't have access to store version (as it is blocked by company's IT policy in my case).

In case you use your own laptop, just find a permissive internet (a friend's maybe). If it is a company PC/laptop and you are in constant use of WSL, especially for docker containers, you need to contact your administrators to have it installed. otherwise, any other way you install it may seem offensive to them.
However, if you are new to WSL and you don't care about `initd`, which is the main problem we were discussing, you can still try installing the internal one, enabling it from windows features. from "app & features" settings, find "related settings" down the list, go to "programs and features" and use "turn windows features on or off" link on the left. it is still a full-featured WSL except for some new features added in recent months. only when you need these new things, then you need the new store version.

**Comment by TechNov** [comment]
if you use WSL. 
run service service_name start insted of ystemctl start service_name

<html>
<body>


Systemd command => Sysvinit command
-- | --
systemctl start service_name=>service service_name start
systemctl stop service_name=> service service_name stop
systemctl restart service_name=> service service_name restart
systemctl status service_name=>service service_name status
systemctl enable service_name => chkconfig service_name on
systemctl disable service_name=> chkconfig service_na


</body>
</html>

**Comment by dieguiariel** [comment]
Hi! im trying to make systemD work on WSL, followed all the instructions from above (i couldn't install KB5020030 as it keeps failing telling me  This Update Is Not Applicable To Your Computer, tried from 21H1 version and upgraded to 22H2 version with no sucess).
Currently:
C:\Users\diegu>wsl --version
Versión de WSL: 1.1.3.0
Versión de kernel: 5.15.90.1
Versión de WSLg: 1.0.49
Versión de MSRDC: 1.2.3770
Versión de Direct3D: 1.608.2-61064218
Versión DXCore: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Versión de Windows: 10.0.19045.2673

Tried with ubuntu 20.04 and upgraded to 22.04 but still didn't work:
 cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04.2 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.2 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy

systemctl status --no-pager
System has not been booted with systemd as init system (PID 1). Can't operate.
Failed to connect to bus: Host is down

sudo cat /etc/wsl.conf
[boot]
systemd=true

any ideas?

**Comment by yilmazdurmaz** [comment]
your wsl version is 1.1.3.0 which means you already have a compatible version. that KB update possibly fails since you try to use arm64 on x64, or vice versa (you downloaded a wrong version). 

> Tried with ubuntu 20.04 and upgraded to 22.04 but still didn't work
each distro is a new virtual machine, and each machine must be configured separately to start with systemd, then must be restarted.
````
wsl --list
wsl -d distronameyouset
sudo nano /etc/wsl.conf # or any editor
exit
wsl -d distronameyouset --shutdown
wsl -d distronameyouset
````

there is one thing I am not sure about: this might require an x64 machine but you use x86 (32bit). and one more possible thing is, I don't know if has any relation, is WSL versus WSL2; can't say if 1.1.3.0 is possible with WSL but worth checking. try the above way first, then check these possibilities.

**Comment by yilmazdurmaz** [comment]
> About the wsl or wsl2 version you pointed me in the right direction:
> wsl -l -v
>   Ubuntu-20.04 Running 1
> wsl --set-version Ubuntu-20.04 2

nice! I wasn't sure about that. It is good to know WLS2 is required to get a working SystemD.

**Comment by ccposhing** [comment]
hi there,
i also got this problem. here is some infomation about in my computer 

`
PS C:\Users\Administrator> wsl -l -v
  NAME               STATE           VERSION
* OracleLinux_8_5    Running         2
PS C:\Users\Administrator> wsl --list
Windows Subsystem for Linux Distributions:
OracleLinux_8_5 (Default)
PS C:\Users\Administrator> wsl --version
WSL version: 1.1.3.0
Kernel version: 5.15.90.1
WSLg version: 1.0.49
MSRDC version: 1.2.3770
Direct3D version: 1.608.2-61064218
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.22621.1344
PS C:\Users\Administrator>
`
`
[root@DESKTOP-2TRGJ9G ~]# cat /etc/wsl.conf
[boot]
systemd=true

[root@DESKTOP-2TRGJ9G ~]# systemctl status maridb
System has not been booted with systemd as init system (PID 1). Can't operate.
Failed to connect to bus: Host is down
[root@DESKTOP-2TRGJ9G ~]#
`

**Comment by yilmazdurmaz** [comment]
@ccposhing , I am guessing you forgot to stop/restart your distro. the last 3 steps in the console:
````
wsl --list -v # need version 2
wsl -d distronameyouset
sudo nano /etc/wsl.conf # or any editor
exit
wsl -d distronameyouset --shutdown
wsl -d distronameyouset
````

**Comment by ccposhing** [comment]
> @ccposhing , your system seems fit. I am guessing you forgot to stop/restart your distro. the last 3 steps in the console:
> 
> ```
> wsl --list -v # need version 2
> wsl --set-version distronameyouset 2
> wsl -d distronameyouset
> sudo nano /etc/wsl.conf # or any editor
> exit
> wsl --terminate distronameyouset # shutdown single
> # wsl --shutdown # shutdown all
> wsl -d distronameyouset
> ```

here is my operation, you can see the restart action, it's still not work 
```
PS C:\Users\Administrator> wsl --terminate OracleLinux_8_5
The operation completed successfully.
PS C:\Users\Administrator> wsl --shutdown
PS C:\Users\Administrator> wsl -d OracleLinux_8_5
[root@DESKTOP-2TRGJ9G Administrator]# systemctl
System has not been booted with systemd as init system (PID 1). Can't operate.
Failed to connect to bus: Host is down
[root@DESKTOP-2TRGJ9G Administrator]# cat /etc/wsl.conf
[boot]
systemd=true
[root@DESKTOP-2TRGJ9G Administrator]#
```

**Comment by yilmazdurmaz** [comment]
@ccposhing , try another distro&version to see if your current one is broken or not. I tried Ubuntu and Oracle on my side and they worked fine.

**Comment by ccposhing** [comment]
@yilmazdurmaz  thanks for help. its caused by the `wsl.conf` file encoding or right i dont konw, but i was use explorer.exe to create this file in windows env before.  when i drop this file and recreate it in wsl bash, it's worked now.😊

**Comment by yilmazdurmaz** [comment]
LOL, new info then: do not use an editor that saves CRLF, or "0A0D" aka dos/windows new line, for line endings :) 

I tested this by using `:set fileformat=dos` in vim and just removing and recreating line endings, then restarted WSL to immediately get this error. I had no idea this would be such a big issue.

**Comment by B0ba-Yaga** [comment]
I tried all the above solutions in Debian the error is the same as the first thread(System has not been booted with systemd as init system (PID 1). Can't operate.
Failed to connect to bus: Host is down)  and in Ubuntu i'm getting (Failed to list units: Transport endpoint is not connected)

Here's my spec:
C:\Users\Sandy>wsl --list -v
  NAME      STATE           VERSION
* Ubuntu    Stopped         2
  Debian    Stopped         2

`C:\Users\Sandy>wsl--version
WSL version: 1.1.3.0
Kernel version: 5.15.90.1
WSLg version: 1.0.49
MSRDC version: 1.2.3770
Direct3D version: 1.608.2-61064218
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.19045.2604

I believe it passes all the required parameters
I have tried doing the above mentioned process several times also restarted Windows and tried it but it's still the same

**Comment by elsaco** [comment]
@B0ba-Yaga on WSL Debian image there's a package missing preventing systemd to work. Install `libpam-systemd` and restart. This was tested with a fresh `wsl --install Debian`

**Comment by yilmazdurmaz** [comment]
Hi @B0ba-Yaga, it seems, in Debian, it might be one of these: either you forgot to create `/etc/wsl.conf` with the correct content, or you created it with an editor that uses CRLF line endings (we discussed just above). I suggest you check these probabilities. however, for Ubuntu, you somehow disabled `systemd` for USER.

I already have Ubuntu 20 and 22 with no issue running it. I have now downloaded the current Debian (update note says 11.6) and with the correct config file, it gave me only the error you described for the Ubuntu: `Failed to connect to bus: No such file or directory`. despite the error, `systemd` is running and the root user can use it.

I searched for this second error, and it is directly related to `DBUS`, in fact to its absence. `apt list --installed | grep bus` will tell you it is missing. I installed dbus for debian, `sudo apt-get install dbus`, and immediately, the problem is solved 🎉  🎉  🎉  (yours should also be solved, I am hoping)

---
Fact summary, in addition to `/etc/wsl.conf` with correct content:
- we need the new WSL 1.x.x series, or Store version (latest 1.1.3.0, as of now),
- and version 2 (WSL2 not WSL), 
- and LF/0A line endings (not CRLF/0D0A)
- also `dbus` inside the distro so that users can use it too. (at least, Debian comes without dbus)

**Comment by yilmazdurmaz** [comment]
I somehow mis-read the error for Ubuntu :) `Failed to list units: Transport endpoint is not connected` is something different that indicates the existence of broken things. please first be the root with `sudo su` and then try `systemctl status` there. if you get the same error, try another Ubuntu distro to check if it is the distro or WSL that is broken. 20.04 and 22.04 are both working fine for me. Depending on what the results are, you may also want to check your custom service files.

**Comment by yilmazdurmaz** [comment]
> @B0ba-Yaga on WSL Debian image there's a package missing preventing systemd to work. Install `libpam-systemd` and restart. This was tested with a fresh `wsl --install Debian`

this library depends on the `dbus`, so installs it too ;) but you do not need this whole packaged deal unless/until you need PAM.

**Comment by elsaco** [comment]
@yilmazdurmaz that's correct, Debian distro (at least the one pulled by `wsl --install Debian` is from the `winget` store, didn't try with the `msstore` version) is lacking `dbus` package, preventing `systemd` from working. After adding it `systemd status --user` is still not working showing `Failed to connect to bus` because of the missing PAM module. After installing `libpam-systemd` it works. Without this plugin(s) PAM won't be able to `Register user sessions in systemd control group hierarchy`. Use `pam-auth-update` for more details.

Quick test:

No `libpam-systemd` package:
```
elsaco@RIPPER:~$ systemctl status --user
Failed to connect to bus: No such file or directory
```

`libpam-systemd` installed:
```
elsaco@RIPPER:~$ systemctl status --user
● RIPPER
    State: running
     Jobs: 0 queued
   Failed: 0 units
    Since: Tue 2023-03-07 11:00:44 PST; 11s ago
   CGroup: /user.slice/user-1000.slice/user@1000.service
           └─init.scope
             ├─139 /lib/systemd/systemd --user
             └─140 (sd-pam)
```

**Comment by yilmazdurmaz** [comment]
@elsaco store version seems to be the same. nice catch 😉 

I am not that much expert, so I shall leave details to pros and make a summary at hand :) in short, we can say, for Debian, systemd works for root, users need `dbus` minimum and `libpam-systemd` for more. And for similar errors other than the "host is down", checking installed packages for `dbus` should be the first step (it will be installed by dependent libraries, or none would be there).

**Comment by B0ba-Yaga** [comment]
> I somehow mis-read the error for Ubuntu :) `Failed to list units: Transport endpoint is not connected` is something different that indicates the existence of broken things. please first be the root with `sudo su` and then try `systemctl status` there. if you get the same error, try another Ubuntu distro to check if it is the distro or WSL that is broken. 20.04 and 22.04 are both working fine for me. Depending on what the results are, you may also want to check your custom service files.
> 
> > @B0ba-Yaga on WSL Debian image there's a package missing preventing systemd to work. Install `libpam-systemd` and restart. This was tested with a fresh `wsl --install Debian`
> 
> this library depends on the `dbus`, so installs it too ;) but you do not need this whole packaged deal unless/until you need PAM.

@yilmazdurmaz .It's still popping the same error with su priviliges for Ubuntu20 and Ubuntu 22 and i don't know why when i run wsl with the changes to wsl.conf it's running extremely slow

**Comment by yilmazdurmaz** [comment]
> Windows version: 10.0.19045.2604

This 19045 is the version number for 22H2. I still use 21H2 which is 19044.

May [this issue @B0ba-Yaga is having](https://github.com/microsoft/WSL/issues/8883#issuecomment-1458492290) be related to 22H2? Can someone test on 22H2 and comment on it?

**Comment by yilmazdurmaz** [comment]
Alright, it is bugging my mind, so I dug a few more. there are issues posted with a similar error `Transport endpoint is not connected` and some users say it might be related to `acpi` which should not be activating for WSL. I have checked  #8867 and #9029.

run `systemctl list-unit-files | grep acpi` and see if you have a service for it. disable systemd if necessary. if so, then there are two possible solutions to try [here](https://github.com/microsoft/WSL/issues/8867#issuecomment-1256344696) and [here](https://github.com/microsoft/WSL/issues/9029#issuecomment-1326907692). if `systemctl` do not work, run `apt list --installed |grep acpi` and maybe try checking for `/usr/lib/systemd/system/acpid.service` or any acpid file nearby under systemd folder.

I am not deep into the solving but they basically say to remove it. worth to check. Please share if this is the case and the issue is solved. this might 

PS: for comparison, my Ubuntu distros (no customization) under W10 21H2 do not have any acpi service, nor acpi installed.

**Comment by crqx3mqzo2** [comment]
hello. I have the same problem.
These are my settings:
powerShell:
PS C:\WINDOWS\system32> wsl --version
Versión de WSL: 1.1.3.0
Versión de kernel: 5.15.90.1
Versión de WSLg: 1.0.49
Versión de MSRDC: 1.2.3770
Versión de Direct3D: 1.608.2-61064218
Versión DXCore: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Versión de Windows: 10.0.19045.2604

NAME                                   FRIENDLY NAME
Ubuntu                                 Ubuntu
Ubuntu-20.04                           Ubuntu 20.04 LTS
Ubuntu-22.04                           Ubuntu 22.04 LTS

distribution:
alfonso@DESKTOP-GUQTUAK:~$ sudo cat /etc/wsl.conf
systemd=true

**Comment by yilmazdurmaz** [comment]
@crqx3mqzo2, `/etc/wsl.conf` must have at least `systemd=true` under `[boot]` section, like this:
````
[boot]
systemd=true
````
your output show you are missing the first line. correct this first, restart, and see if you have other problems. then check other answers I provided here about important things to note, such as CRLF line endings (I summarize/edit them whenever possible). When all fails, please tell us again.

**Comment by crqx3mqzo2** [comment]
Oh, I understand.  thank you

**Comment by xkex36** [comment]
> it is not `%userprofile%\.wslconfig` you would be editing. it is `/etc/wsl.config` inside. you will need to have root access
> 
> ```
> wsl  # in cmd or powershell
> sudo nano /etc/wsl.conf # or use your fav editor
>   #insert those 2 lines you have
> exit
> wsl --shutdown #  first shutdown
> wsl # then restart
> ```
> 
> PS: this `.wslconfig` thing might be something that wsl team have in mind. at least for now, that is the other file you need to edit.

bro how you execute or exite

**Comment by yilmazdurmaz** [comment]
@xkex36  you need to run these command in terminal opened by `cmd`, "powershell", "pwsh", "Windows Terminal" or any windows terminal, except the one opened by the `wsl` command. (editing is done inside wsl terminal)

**Comment by xkex36** [comment]
im sorry i still dont get it . im gonna tell you what i did : 
- i openned powershell as admin
- then i run wsl
- then this command sudo nano /etc/wsl.conf
- then i wrote the two lines
-
![Screenshot (36)](https://user-images.githubusercontent.com/127667483/224578104-1b137485-f43f-48b2-9c96-ee4e5d8771ed.png)
then i didnt know what to do next

**Comment by yilmazdurmaz** [comment]
Oh, Ok, a newbie to Linux command-line editors (or as a whole for such editors) :) you should say so first. The only thing missing in my instruction is "save file and exit editor" step.

For a windows user, these editors and their difficulty levels are an easy thing to forget when instructing. one will need to learn at least one of these editors. they are heavily dependent on keyboard shortcuts to do things. Strive for information for any editor on how to do 3 steps; edit, save, and exit, because they may all use different rituals for the purpose. Many of them use compatible keys to GUI editors, but others are way too off from GUI counterparts.

`nano` is one of the easiest to use. you open a file, write content, "Ctrl+O" to save while writing, "Ctrl+X" to begin exit (not immediate exit), decide to save or not (select no if messed), accept file name (give new name if started without one), and that is it.

I recently started using `vi/vim` again* (nightmare for beginners, weapon for masters). it comes by default in almost any Linux distro (unless lightweight like a container). Yet, unlike others, it starts in shortcut mode, so do not ever try typing before learning how to exit it, what the colon, ":", do, and why "ESC" is your best friend.

**Comment by B0ba-Yaga** [comment]
> Alright, it is bugging my mind, so I dug a few more. there are issues posted with a similar error `Transport endpoint is not connected` and some users say it might be related to `acpi` which should not be activating for WSL. I have checked #8867 and #9029.
> 
> run `systemctl list-unit-files | grep acpi` and see if you have a service for it. disable systemd if necessary. if so, then there are two possible solutions to try [here](https://github.com/microsoft/WSL/issues/8867#issuecomment-1256344696) and [here](https://github.com/microsoft/WSL/issues/9029#issuecomment-1326907692). if `systemctl` do not work, run `apt list --installed |grep acpi` and maybe try checking for `/usr/lib/systemd/system/acpid.service` or any acpid file nearby under systemd folder.
> 
> I am not deep into the solving but they basically say to remove it. worth to check. Please share if this is the case and if the issue is solved.
> 
> PS: for comparison, my Ubuntu distros (no customization) under W10 21H2 do not have any acpi service, nor acpi installed.

@yilmazdurmaz Thanks alot as described in #9029 removing acpid has worked for me

**Comment by fredxia** [comment]
On Windows 10, running these commands fixed systemd issue:

```
sudo apt-get update && sudo apt-get install -yqq daemonize dbus-user-session fontconfig

sudo daemonize /usr/bin/unshare --fork --pid --mount-proc /lib/systemd/systemd --system-unit=basic.target
```

Found it from this link: https://gist.github.com/alyleite/ca8b10581dbecd722d9dcc35b50d9b2b

**Comment by yilmazdurmaz** [comment]
@fredxia , using `daemonize` is just overkill as it is an ancient relic of programs. and at least for Ubuntu and Debian distros, there is no need for such a thing either, let alone if it works as it should for this is not a stand-alone Linux system because, again, it was created in ancient times before Linux even existed. [ubuntu change logs](https://packages.ubuntu.com/focal/admin/daemonize) for example has "unstable" in each version maintained last 10 years.

However here, there are just a couple of missing things in instructions, and this discussion has them most, if not all to solve problems in modern (and easy) ways. From the command you gave, it seems that system is possibly a Debian and it misses `dbus`. just installing it would even solve their problem.
[here in this answer above from @elsaco ]https://github.com/microsoft/WSL/issues/8883#issuecomment-1458687389) tells `libpam-systemd` would solve without going fancy.

**Comment by fredxia** [comment]
@yilmazdurmaz, I'm on Ubuntu 22.04. Just install dbus does not work for me. I have done the config setup steps discussed in comments above. Unfortunately they don't work. There is probably other better ways. But this is the solution I got so far.

**Comment by yilmazdurmaz** [comment]
@fredxia I would like to hear about your setup because Ubuntu comes with dbus related packages and I did not have such problem with it before. I try to expand my answers here whenever we have a new problem, but they are mainly caused by misinterpretation of documentation. Actually, for ubuntu, steps work almost out-of-the-box. I suggest you, for future installations, check my comments here and see where you (or your system) deviated to give you that error because `daemonize` can cause more harm later in your workflow as (again) its change logs state it to be unstable.

**Comment by fredxia** [comment]
@yilmazdurmaz, daemonize is not necessary. It's just a simple way to free the terminal. But rest of command line is needed to for me start systemd in WSL2.

**Comment by yilmazdurmaz** [comment]
you are not helping. what you gave as a solution is just a workaround.  help us identify the problem in your system that you could not see. and again, Ubuntu 22.04 is proven to work without any fanciness.

for the sake of progress, first check this one: https://github.com/microsoft/WSL/issues/8883#issuecomment-1345472665

then check and give outputs for these commands.
````
wsl--version           # cmd or powershell, 1.0.0.0 or higher "store" version
wsl --list -v          # cmd or powershell, WSL2 is version 2
cat /etc/wsl.conf     # inside the distro shell, need correct content (text viewer)
hexdump /etc/wsl.conf  # inside distro shell, 0d0a or 0a? (hex viewer)
````
and, give the Ubuntu installation commands and a repo address if you did not install it from the Store.

**Comment by edemir206** [comment]
> @fredxia , using `daemonize` is just overkill as it is an ancient relic of programs. and at least for Ubuntu and Debian distros, there is no need for such a thing either, let alone if it works as it should for this is not a stand-alone Linux system because, again, it was created in ancient times before Linux even existed. [ubuntu change logs](https://packages.ubuntu.com/focal/admin/daemonize) for example has "unstable" in each version maintained last 10 years.
> 
> However here, there are just a couple of missing things in instructions, and this discussion has them most, if not all to solve problems in modern (and easy) ways. From the command you gave, it seems that system is possibly a Debian and it misses `dbus`. just installing it would even solve their problem. [here in this answer above from @elsaco ][#8883 (comment)](https://github.com/microsoft/WSL/issues/8883#issuecomment-1458687389)) tells `libpam-systemd` would solve without going fancy.

Yes!!!In order to have systemd working with Debian image I had to do the following:

1)  `wsl --install Debian`

2) `sudo cat /etc/wsl.conf`
Add the following content: 
`[boot]
systemd=true`

3) command: `sudo apt install libpam-systemd `

4) In powershell run `wsl --shutdown`

5) Enter your WSL distro again and check if systemd is running with this command: systemctl --no-pager status user.slice > /dev/null 2>&1 && echo 'OK: Systemd is running' || echo 'FAIL: Systemd not running'

If it says **OK: Systemd is running** everything is running fine.

**Comment by globalistas** [comment]
> 2\. `sudo nano /etc/wsl.conf`
>     Add the following content:
>     `[boot] systemd=true`

The file doesn't exist on my Debian system and after creation and following your steps my systemd is still not running in WSL.

**Comment by edemir206** [comment]
> 

Thanks, just fixed it.

**Comment by Hormozzan** [comment]
@dieguiariel 
Hola, dear Diegu. Your solution worked for me.
Muchas gracias🙏

**Comment by digital-marketing-engineer** [comment]
Thanks, This https://github.com/microsoft/WSL/issues/8883#issuecomment-1455090207 worked for me

**Comment by AlphaMAtrix69** [comment]
I have been trying to edit wsl.conf file but it is giving error
sudo nano /etc/wsl.conf
-bash: sudo: command not found

**Comment by yilmazdurmaz** [comment]
> I have been trying to edit wsl.conf file but it is giving error
> wslg [ /etc ]$ sudo nano /etc/wsl.conf
> -bash: sudo: command not found
assuming you are running the command "inside" the distro's terminal, your distro seems not to have it by default. check its documentation on how to run root-privileged commands, or how to install sudo command. maybe it does not even need it. (also check if nano is installed, and install it or use another editor you already installed inside)

**Comment by jlowe000** [comment]
I found this thread useful. I was struggling with the upgrade and realised that the WSL2 was installed but I also needed to update existing distros to be of the new version. I found this comment and the following one very useful. https://github.com/microsoft/WSL/issues/8883#issuecomment-1455090207

**Comment by rezakusuma40** [comment]
hi all, I've got the same problem, it seems I have done all the isntruction but still no result
here's my setting:
i run these using cmd
C:\Users\msi>wsl --status
Default Distribution: Ubuntu
Default Version: 2

Windows Subsystem for Linux was last updated on 05/04/2023
The Windows Subsystem for Linux kernel can be manually updated with 'wsl --update', but automatic updates cannot occur due to your system settings.
To receive automatic kernel updates, please enable the Windows Update setting: 'Receive updates for other Microsoft products when you update Windows'.
For more information please visit https://aka.ms/wsl2kernel.

Kernel version: 5.10.16

C:\Users\msi>wsl --version #for some unknown reason it's invalid
Invalid command line option: --version

I am using windows 10 version 22h2, x64. I tried to install relevant KB5020030 but it says the updater is not applicable on my computer

i run this line using wsl, it's already correct
myusername@DESKTOP-LPKC9LD:~$ cat /etc/wsl.conf
[boot]
systemd=true

but still...
myusername@DESKTOP-LPKC9LD:~$ systemctl status --no-pager
System has not been booted with systemd as init system (PID 1). Can't operate.
Failed to connect to bus: Host is down

**Comment by yilmazdurmaz** [comment]
> C:\Users\msi>wsl --version #for some unknown reason it's invalid
> Invalid command line option: --version

Hi @rezakusuma40,  I think you are running the default wsl version. you need the "store" version.

**Comment by elsaco** [comment]
@rezakusuma40 `systemd` official support was added to WSL release 0.67.6. You need this version or newer to enable systemd. See details at https://github.com/microsoft/WSL/releases/tag/0.67.6

**Comment by rezakusuma40** [comment]
@yilmazdurmaz @elsaco I have checked the link above, and it turns out that I just need to type "wsl --update" in the command prompt to update to the store version. and it works after that, thank you

**Comment by lucliu20** [comment]
Hello all,
The similar issue occurs to me, below is my system info.

_wsl --version_
WSL version: 1.2.5.0
Kernel version: 5.15.90.1
WSLg version: 1.0.51
MSRDC version: 1.2.3770
Direct3D version: 1.608.2-61064218
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.19044.3086

sudo cat _/etc/wsl.conf_
[boot]
systemd=true


_reboot_
System has not been booted with systemd as init system (PID 1). Can't operate.
Failed to connect to bus: Host is down

**Comment by yilmazdurmaz** [comment]
@lucliu20, use a hex viewer on that file (such as "hexdump") and see if it uses LF/0A line endings (not CRLF/0D0A). this is the most possible cause when everything seems: caused by missing things inside the distro. (also dbus and acpi, too)

if you see 0D0A, delete the file and remake it within the distro with a linux editor (such as "nano")

**Comment by DexSpecial** [comment]
> /etc/wsl.conf

Permissions denied.

**Comment by yilmazdurmaz** [comment]
@DexSpecial , add "sudo" to your commands when you see this. "/etc" is one of those protected folders and changes in it need admin rights. (we sometimes forget to mention this for newcomers)
`sudo nano /etc/wsl.conf`

**Comment by igwen6w** [comment]
Hello all,
The similar issue occurs to me, below is my system info. I tried to install relevant KB5020030 but it says the updater is not applicable on my computer. Any idea? Thanks.

wsl --version
WSL version: 1.2.5.0
Kernel version: 5.15.90.1
WSLg version: 1.0.51
MSRDC version: 1.2.3770
Direct3D version: 1.608.2-61064218
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.19045.3448

sudo cat /etc/wsl.conf
[boot]
systemd=true

hexdump /etc/wsl.conf
0000000 625b 6f6f 5d74 730a 7379 6574 646d 743d
0000010 7572 0a65
0000014

reboot
System has not been booted with systemd as init system (PID 1). Can't operate.
Failed to connect to bus: Host is down

**Comment by yilmazdurmaz** [comment]
@igwen6w  KB5020030  is a now-old update for the old system WSL to update it to store version. Yours is already the store version 1.2.5.0. config file also seem fine with 0x0A line ending and at correct path.

My guess is that the distribution you use has other missing dependencies like dbus or acpid. for example, Debian distros has such issues. Please try to check my answers above as they are not short to tell here in one line. If these don't solve the issue it might be another dependency I am not aware of relating to systemd.

Ubuntu 20 and 22 were fine as they were. if your issue is for them, something might be broken in your installation or something went wrong with this version of WSL. try updating or downgrading.

**Comment by rahberkazmi128** [comment]
@yilmazdurmaz 
I just introduced and started ubuntu in wsl ( windows terminal). everything works fine. Just this happens:

 sudo reboot
[sudo] password for kazmi:
System has not been booted with systemd as init system (PID 1). Can't operate.
Failed to connect to bus: Host is down
Failed to talk to init daemon.

When I look into my wsl, i got 2 reviews:
first in ubuntu windows terminal:
 wsl --version
Unknown command: --version
WSL
Wsman Shell commandLine, version 0.2.1

USAGE: wsl COMMAND [PARAMS...]

COMMANDS:
identify  - WS-Identify
enum      - WS-Enumerate
get       - WS-Get
put       - WS-Put
invoke    - WS-Invoke
xclean    - Delete all files generated by this tool set
xcred     - Create or display credential file
xcert     - Get server certificate (saved to <IPADDRESS>.crt)

PARAMS specification is specific to a COMMAND.

Output will be saved to ./response.xml. If you want to run parallel
executions in the same directory, define RTFILEPREFIX in the environment.
Doing so may significantly increase files generated.

Requires: curl, xmllint, GNU core utilities.
Optional: xsltproc for output formatting, gpg for encrypted credential.
Optional: wget as alternate for curl when not available.

and all wsl commands not working here, likewise wsl --shutdown too.

Second, when I looked into powershell :
wsl --version
WSL version: 1.2.5.0
Kernel version: 5.15.90.1
WSLg version: 1.0.51
MSRDC version: 1.2.3770
Direct3D version: 1.608.2-61064218
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.19045.3516

I just want toknow, how I'm going to resolve this . I tried many sets of instruction . but still . Nothing works.
I also got happened to read this blog.
https://itslinuxfoss.com/fix-system-not-booted-systemd-as-init-system-pidcant-operate/

Please let me help me know what I should do and how.

**Comment by yilmazdurmaz** [comment]
@rahberkazmi128  it seems you somehow messed your environment variables to get these two outputs. The first one you run is not WSL as we speak. You seem to have installed another utility called "Wsman Shell commandLine" and unfortunately it has the exact same command line name "wsl". that is why WSL parameter do not for with it. If you do not use wsman utility as  much as WSL, rename its executable name, or remove it from PATH, or uninstall completely, you choice.

your powershell seems to run the correct WSL command, so please run it there and give the error message from there, if any. make sure you have followed our warnings here and have the correct config file.

**Comment by yilmazdurmaz** [comment]
@rahberkazmi128  it seems you somehow messed your environment variables to get these two outputs. The first one you run is not WSL as we speak. You seem to have installed another utility called "Wsman Shell commandLine" and unfortunately it has the exact same command line name "wsl". That is why WSL parameters do not work with it. If you do not use wsman utility as much as WSL, rename its executable name, or remove it from PATH, or uninstall completely, your choice.

Your powershell seems to run the correct WSL command, so please run it there and give the error message from there, if any. make sure you have followed our warnings here and have the correct config file.

**Comment by igwen6w** [comment]
> @igwen6w KB5020030 is a now-old update for the old system WSL to update it to store version. Yours is already the store version 1.2.5.0. config file also seem fine with 0x0A line ending and at correct path.
> 
> My guess is that the distribution you use has other missing dependencies like dbus or acpid. for example, Debian distros has such issues. Please try to check my answers above as they are not short to tell here in one line. If these don't solve the issue it might be another dependency I am not aware of relating to systemd.
> 
> Ubuntu 20 and 22 were fine as they were. if your issue is for them, something might be broken in your installation or something went wrong with this version of WSL. try updating or downgrading.

Thanks. You're right, reinstalling wsl and ubuntu22.04 fixed the problem.

**Comment by Khun111** [comment]
> > it is `/etc/wsl.config` inside
> 
> @Khun111 start with this line, and please read slowly. it is relatively easy, but there are many things to be careful of. for example, use above commands sequentially from `cmd` or `powershell`, as the above commands says after `#` (as comments). For ubuntu, though, these commands you quoted should be enough. otherwise, check the discussion more thoroughly for other possible problems and solutions.

Oh the file isn't there is what I'm saying but I created a /etc/wsl.conf as a new file and enabled it there. Looks like it works

**Comment by Khun111** [comment]
> > it is `/etc/wsl.config` inside
> 
> @Khun111 start with this line, and please read slowly. it is relatively easy, but there are many things to be careful of. for example, use above commands sequentially from `cmd` or `powershell`, as the above commands says after `#` (as comments). For ubuntu, though, these commands you quoted should be enough. otherwise, check the discussion more thoroughly for other possible problems and solutions.

Oh looks like systemd was available in my distro all along. My issue is prolly trying to start init services with systemd. Says system has not been booted with systemd as init. Although  when I run ps --no-headers -o comm 1 I still get init and not systemd

**Comment by Jon-the-2nd** [comment]
Anyone who comes here looking to enable systemd on kali linux in WSL, please note that enabling systemd will break WSL interop and prevent windows binaries from being executed (including the win kex client):

https://github.com/microsoft/WSL/issues/9887#issuecomment-1516434388

**Comment by cscemail20** [comment]
@dieguiariel Thanks , bro.
wsl --set-version Ubuntu-2204 2

**Comment by lokfan** [comment]
Installed Windows, WSL2 and Ubuntu this weekend. In '/etc/wsl.conf' it was already set to true, but systemd did not work.
Out of frustration I changed it to false, shutdown, change back to true and shutdown again. Now systemd is working :tired_face: 

ps. `wsl --version` returns help..

**Comment by yilmazdurmaz** [comment]
> Installed Windows, WSL2 and Ubuntu this weekend. In '/etc/wsl.conf' it was already set to true, but systemd did not work. Out of frustration I changed it to false, shutdown, change back to true and shutdown again. Now systemd is working 😫

keep in mind: if something does not work right away, try restarting first before thinkering further :) 

> ps. `wsl --version` returns help..

maybe something went off in very latest version, or else maybe your machine also needs a reboot.

**Comment by nathanneibauer** [comment]
This is failing for me today. Because of a system corruption, I had to purge `wsl` completely from my machine and reinstall it. 

The problem: when I try to use systemctl, I get the following:
```
System has not been booted with systemd as init system (PID 1). Can't operate.
Failed to connect to bus: Host is down
```

However, my `/etc/wsl.conf` is already populated with

```
[boot]
systemd=true
```


I have tried restarting wsl and restarting my windows machine, but to no avail. I am running Ubuntu 22.04.3 and wsl kernel version 5.10.16:
![image](https://github.com/microsoft/WSL/assets/116027047/692726ab-51b0-451a-9694-aba9699bc7cb)


![image](https://github.com/microsoft/WSL/assets/116027047/7bd52d08-d9e0-4ad7-bf9f-c23c6c77c029)

**Comment by elsaco** [comment]
@nathanneibauer looks like your WSL was updated on 2/12/2014, but the kernel version is `5.10.16`. That's an old kernel. Latest WSL kernel is `5.15.146.1`. 

I just did a quick test and systemd is working right after installation:

```
Installing: Ubuntu 22.04 LTS
Ubuntu 22.04 LTS has been installed.
Launching Ubuntu 22.04 LTS...
Installing, this may take a few minutes...
Please create a default UNIX user account. The username does not need to match your Windows username.
For more information visit: https://aka.ms/wslusers
Enter new UNIX username: elsaco
New password:
Retype new password:
passwd: password updated successfully
The operation completed successfully.
Installation successful!
To run a command as administrator (user "root"), use "sudo <command>".
See "man sudo_root" for details.

Welcome to Ubuntu 22.04.3 LTS (GNU/Linux 5.15.146.1-microsoft-standard-WSL2 x86_64)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage


This message is shown once a day. To disable it please create the
/home/texas/.hushlogin file.
elsaco@texas:~$ cat /etc/wsl.conf

[boot]
systemd=true
elsaco@texas:~$ systemctl status
● eleven
    State: running
     Jobs: 0 queued
   Failed: 0 units
    Since: Mon 2024-02-12 13:56:45 PST; 16s ago
   CGroup: /
---cut---
```

WSL info:

```
WSL version: 2.1.1.0
Kernel version: 5.15.146.1-2
WSLg version: 1.0.60
MSRDC version: 1.2.5105
Direct3D version: 1.611.1-81528511
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.22631.3085
```

**Comment by yilmazdurmaz** [comment]
Hi @doraemonxxx, Here, this might be the culprit in your case  ;)

> `dism.exe ... /norestart`

Please first restart your PC and then check if the problem continues. Also check if the "default" is to use version 2, not version 1.

https://learn.microsoft.com/en-us/windows/wsl/install#upgrade-version-from-wsl-1-to-wsl-2

Ubuntu 22.04 comes with a config file set for systemd, thus it should work fine if WSL gets installed fine and set correctly.

**Comment by Haitam-Elgharras** [comment]
After spending a day troubleshooting, I resolved the "permission denied" issue with Docker by following these steps:

1. Added my user to the `docker` group:
   ```
   sudo usermod -aG docker $USER
   ```

2. Checked the status of the Docker daemon:
   ```
   sudo systemctl status docker
   ```

3. Started the Docker daemon if it wasn't running:
   ```
   sudo systemctl start docker
   ```

4. Verified the permissions of the Docker socket:
   ```
   ls -l /var/run/docker.sock
   srw-rw---- 1 root docker 0 Jan 24 10:06 /var/run/docker.sock
   ```

5. Adjusted the permissions of the Docker socket if it's not as mentioned:
   ```
   sudo chmod 666 /var/run/docker.sock
   ```

6. Restarted the Docker service:
   ```
   sudo systemctl restart docker
   ```

7. Close the terminal and open it again, then try :
   ```
   docker version
   ```

These steps helped me successfully connect to the Docker daemon without encountering permission issues.

**Comment by sertraline** [comment]
I'm running a fully updated 21H2 version, was unable to start WSL with systemd enabled.  My wsl.conf came with
```
[boot]
systemd=true
```
by default, nevertheless, systemd was not available.

I installed "Windows Subsystem for Linux" from the MS Store, like I always did, which is apparently where the issue comes from. Store version of WSL is broken, and it does not even come up in the Store search (it was hidden and I had to google for the URL).  

`wsl --status`  reported that I indeed use WSL2, but my WSL kernel version was 5.10.102.1. Apparently, this version does not support systemd and WSLg. `wsl --update` also reported that my WSL is updated to the last available version, which is not true.  Latest WSL kernel at this moment is 5.15, but this kernel is not listed on MS's [own update catalog](https://www.catalog.update.microsoft.com/Search.aspx?q=wsl).  

I tried to download latest WSL from the Github releases page, but the installer just crashed.

My solution was to install WSL from dism:
`dism.exe /online /enable-feature /featurename:Microsoft-Windows-Subsystem-Linux /all /norestart`  
After that, reboot and run `wsl --update`. Now my kernel version is 5.15 and systemd works properly.

**Comment by MJQWIND** [comment]
> I'm running a fully updated 21H2 version, was unable to start WSL with systemd enabled. My wsl.conf came with
> 
> ```
> [boot]
> systemd=true
> ```
> 
> by default, nevertheless, systemd was not available.
> 
> I installed "Windows Subsystem for Linux" from the MS Store, like I always did, which is apparently where the issue comes from. Store version of WSL is broken, and it does not even come up in the Store search (it was hidden and I had to google for the URL).
> 
> `wsl --status` reported that I indeed use WSL2, but my WSL kernel version was 5.10.102.1. Apparently, this version does not support systemd and WSLg. `wsl --update` also reported that my WSL is updated to the last available version, which is not true. Latest WSL kernel at this moment is 5.15, but this kernel is not listed on MS's [own update catalog](https://www.catalog.update.microsoft.com/Search.aspx?q=wsl).
> 
> I tried to download latest WSL from the Github releases page, but the installer just crashed.
> 
> My solution was to install WSL from dism: `dism.exe /online /enable-feature /featurename:Microsoft-Windows-Subsystem-Linux /all /norestart` After that, reboot and run `wsl --update`. Now my kernel version is 5.15 and systemd works properly.

Thank you very much! I met the same problem and with your help I solved it.

**Comment by jwhipp** [comment]
+1 here... seeing this now on a fresh Windows 11 install (22631.3527) using "Turn Windows features on or off" and selecting "Windows subsystem for Linux".   I used the dism.exe trick above with apparently updates it again with a newer version and sovles the issue.  My understanding has always been that dism and "Turn Windows features ..." were basically the same thing.

**Comment by ndamulelonemakh** [comment]
I can confirm, adding:

```bash
[boot]
systemd=true
```

inside `/etc/wsl.conf` worked for me on **Windows version: 10.0.19045.4291** running **WSL version: 2.0.14.0**

**Comment by owlcode3** [comment]
I'm facing similar issue and I've tried everything suggeste here, still can't get it to work.

**Comment by suyash2819** [comment]
I am still not able to resolve:

WSL version: 2.2.4.0
Kernel version: 5.15.153.1-2
WSLg version: 1.0.61
MSRDC version: 1.2.5326
Direct3D version: 1.611.1-81528511
DXCore version: 10.0.26091.1-240325-1447.ge-release
Windows version: 10.0.19045.4529

added

`[boot]
systemd=true`

in /etc/wsl.conf

**Comment by yilmazdurmaz** [comment]
@owlcode3  "tried everything" is a very misleading statement. you possibly missed one of the working solutions we discussed above. this is a long thread so please make sure you followed messages correctly, and if it still fails prepare a more detailed "help me" post, that includes WSL version, config file path and content etc.

@suyash2819 did you copy-pasted as-is? you got a single line here, but it sould be at 2 lines (I am sure you have already seen how).

for both of you, please make sure you restarted WSL at leat, or you PC to make sure, after you installed it clean.

**Comment by owlcode3** [comment]
> @owlcode3 "tried everything" is a very misleading statement. you possibly missed one of the working solutions we discussed above. this is a long thread so please make sure you followed messages correctly, and if it still fails prepare a more detailed "help me" post, that includes WSL version, config file path and content etc.
> 
> @suyash2819 did you copy-pasted as-is? you got a single line here, but it sould be at 2 lines (I am sure you have already seen how).
> 
> for both of you, please make sure you restarted WSL at leat, or you PC to make sure, after you installed it clean.

How'd you tagged my comment as misleading when you don't know all I've done to troubleshoot the issue?
Well I opened an issue [here](https://github.com/microsoft/WSL/issues/11690#issuecomment-2168382460) with logs also. You can check it out if you don't mind.

**Comment by yilmazdurmaz** [comment]
> How'd you tagged my comment as misleading when you don't know all I've done to troubleshoot the issue?

because everyone tell "I have tried everything", yet they do not tell us what they have really tried, and eventually we do find out that they haven't really tried everything because they have really missed something. And here you are telling us the same thing without giving us any details on what you have tried. 

Please, stop arguing and enlighten us in that issue you have created. I am in the hospital for a family member and don't have much time, so I may not be the one to give you a solution. yet, i will try to check your logs later. so please keep your cool and patience.

**Comment by suyash2819** [comment]
@yilmazdurmaz I have written it in 2 lines, have shutdown the WSL and restarted - it didn't work, and I restarted my PC but still it's the same.

![image](https://github.com/microsoft/WSL/assets/28905722/982792ef-66ea-426b-a4fb-5f28981765c3)

do you think the config which I shared from WSL --version are correct?

**Comment by dieguiariel** [comment]
Hi! just to check, execute from windows command line: *wsl --list -v*  to
check if your vm is using wsl version 2.

Regards.

**Comment by suyash2819** [comment]
I uninstalled, reinstalled, it gave another error which I had to correct from my BIOS and then it all worked.
Thank you

**Comment by owlcode3** [comment]
> I uninstalled, reinstalled, it gave another error which I had to correct from my BIOS and then it all worked. Thank you

How were you able to do it?

**Comment by owlcode3** [comment]
> wsl --update



> I'm running a fully updated 21H2 version, was unable to start WSL with systemd enabled. My wsl.conf came with
> 
> ```
> [boot]
> systemd=true
> ```
> 
> by default, nevertheless, systemd was not available.
> 
> I installed "Windows Subsystem for Linux" from the MS Store, like I always did, which is apparently where the issue comes from. Store version of WSL is broken, and it does not even come up in the Store search (it was hidden and I had to google for the URL).
> 
> `wsl --status` reported that I indeed use WSL2, but my WSL kernel version was 5.10.102.1. Apparently, this version does not support systemd and WSLg. `wsl --update` also reported that my WSL is updated to the last available version, which is not true. Latest WSL kernel at this moment is 5.15, but this kernel is not listed on MS's [own update catalog](https://www.catalog.update.microsoft.com/Search.aspx?q=wsl).
> 
> I tried to download latest WSL from the Github releases page, but the installer just crashed.
> 
> My solution was to install WSL from dism: `dism.exe /online /enable-feature /featurename:Microsoft-Windows-Subsystem-Linux /all /norestart` After that, reboot and run `wsl --update`. Now my kernel version is 5.15 and systemd works properly.

I did exactly what you stated here. for some reasons my kernel version still remains at 5.10.102.1

**Comment by yilmazdurmaz** [comment]
@owlcode3, I could not follow changes to WSL in recent months. It is possible 21H2 got old and current releases might not work on it. Small possibility, nonetheless a possibility

Other than that, you might have enabled Hyper-V on your first install and it might be the reason yours don't update to newer versions. Check if this is the case and disable its traces.

Also try to completely remove all WSL features along with Hyper-V. Then try again this "dism" command.

I don't think it would be a thing for you, but also check your BIOS if the virtualization is enabled.

**Comment by owlcode3** [comment]
> @owlcode3, I could not follow changes to WSL in recent months. It is possible 21H2 got old and current releases might not work/install on it. Small possibility, nonetheless a possibility
> 
> Other than that, you might have enabled Hyper-V on your first install and it might be the reason yours don't update to newer versions. Check if this is the case and disable its traces.
> 
> Also try to completely remove all WSL features along with Hyper-V. Then try again this "dism" command.
> 
> I don't think it would be a thing for you, but also check your BIOS if the virtualization is enabled.

I think It's called HypervisorPlatform on my system, and I've enabled and disabled It, It doesn't make any difference.
And yes, virtualization is enabled.

**Comment by MeechaPooch** [comment]
Its still not working
Ive added the /etc/wsl.conf, ive rebooted wsl and my dell computer, my WSL is at the latest version. I am on windows 10 home edition.

Error print:
```
$ systemctl
System has not been booted with systemd as init system (PID 1). Can't operate.
Failed to connect to bus: Host is down
```

/etc/wsl.conf:
```
$ cat /etc/wsl.conf
[boot]
systemd=true

[user]
default=powchm
```

latest version 2.2.4.0 and windows version info
```
$ wsl.exe --version
Default Distribution: Ubuntu
Default Version: 2
PS C:\Users\Irene> wsl --version
WSL version: 2.2.4.0
Kernel version: 5.15.153.1-2
WSLg version: 1.0.61
MSRDC version: 1.2.5326
Direct3D version: 1.611.1-81528511
DXCore version: 10.0.26091.1-240325-1447.ge-release
Windows version: 10.0.19045.4529
```

**Comment by yilmazdurmaz** [comment]
@MeechaPooch  you seem you have edited the config file to add default user. My guess is that you used an editor that changed the line endings from linux format (LF-0A) to dos format (CRLF-0D0A). system fails to load the config file in this case.

try `hexdump /etc/wsl.conf` and see if this is the case (does Ubuntu have hexdump by default? you may need to install it)

Fixing it depends on the editors to use and their settings for this format, and it will be long to give instructions here. Please search the internet about fixing this line ending problem.

**Comment by MeechaPooch** [comment]
I've verified that it is properly loading the conf file and still not booting with systemd

**Comment by yilmazdurmaz** [comment]
> I've verified that it is properly loading the conf file and still not booting with systemd

I am clueless now. Your system seems ready and you say the config file is fine.

I wonder if something has changed (or a bug) about the "Home" editions. Do you guys have a chance to install from scratch including Windows? (chance to use another version?)

**Comment by yilmazdurmaz** [comment]
I had a recent installation experience, and honestly it should be this easy now. My old pc/laptop all broke and I was on a borrowed Mac for a while until I found out to side install windows on it. Then had to try my old WSL disks since the disk they are on was giving errors. so I had to install WSL again.

https://learn.microsoft.com/en-us/windows/wsl/install-manual#step-4---download-the-linux-kernel-update-package

The link has a recent (from last december) steps to install WSL 2. First 3 steps are there to make sure WSL is installed. Then, on step 4, after running that kernel update package, I used the 'wsl --update' command and then 'wsl --version' command worked giving the current 2.2 number. (don't forget to step 5 after this)

after importing one of my old images, it was running fine with systemd as the PID 1, showing these 4-5 steps are nowadays pretty enough to run systemd, unless your linux distro or config file has the problems I listed in my early answers.

**Comment by yilmazdurmaz** [comment]
I want to share my experience about why it might be stuck at ``5.10.xx`` version line. It will be bit long story and I hope its steps may help newcomers to identify their problems. (jump to last "here are the steps" part if you dont like reading stories)

I had a few days to work on the install/update steps. (I messed up my registry keys, so had to reinstall). I have found out that I was just lucky not get any problem because of one of my usual system installation step: install system updates then do whatever else needed.

This time around, I tried to install WSL/2 before any update: Yep, it totally got me unprepared, and messed with my sanity. As I had never have a problem getting the latest version before, it was though for me to identify the problem. That small update file from step 4 on the above link was installed, yet the version was still the old one.

On Win 10 22H2, the default version that comes is ``5.10.16``. And if I use `wsl --update`, it gets ``5.10.102.1``. I used `--rollback` option along with trying to uninstall/reinstall from windows features, and messed up more: it crippled so much that it wouldn't even uninstall (then I reinstall windows again, as it was not clean enough for me anymore).

In the next clean attempt, I have found the store link to the WSL/2 update (is missing from store nowadays) and installed it. Surprise!! It did not work as expected. Even after installing store version, it kept the ``5.10.xx`` line. The wsl link/program on the start menu was just not running at all.

Finally I went a little deeper. I have fount out that the new wsl version is installed under ``c:\Program Files\WSL\`` and the ``wsl.exe`` in there had the ``2.2.xx`` version. It is just that it wouldn't work if I click on it. Then I decided to run it in the console to see if it was giving any error, and Voila!!, it did:

> Windows version 10.0.19045.2006 does not support the packaged version of Windows Subsystem for Linux.
> Install the required update via Windows update or via: https://aka.ms/store-wsl-kb-win10
> For information please visit https://aka.ms/wslinstall
> Error code: Wsl/WSL_E_OS_NOT_SUPPORTED

I followed the link and found out that I had already posted about it, alas without knowing why exactly: At least update your system using ``KB5020030`` if you clean install your system.

As soon as I installed this update, my wsl was running the latest `2.2.xx` version, along with the ``systemd`` being PID 1 in my ubuntu 22.04. (previously it was starting with ``initd`` being PID 1).

I reinstalled windows one more time and tried a slightly different approach: installed all windows updates that the Windows Update in the settings finds, and use only a single command that comes with default wsl, `wsl --update`. It worked like a charm and without downloading anything else, it updated itself to the latest store version (previously this command was updating within 5.10.xx line)

Long story short, here are the 4 steps to install ``WSL 2`` in 2024 (first one is very important):
- Step 1: Update your Windows. 
  - the WSL that comes with installation media cannot update itself beyond ``5.10.xx`` version line.
  - You need to have at least one of ``1904x.2311`` version of Windows 10 (20H2, 21H1, 21H2, 22H2) that was updated on December 2022 with ``KB5020030``, to run WSL 2. Follow https://aka.ms/store-wsl-kb-win10.
  - Or, just open the Windows Updates tab in the Updates& Security of Settings application, and fully update your system.
  - This is again given in step 3 in the link I gave in my previous post, but I think it fails to deliver the importance of the step.

- Step 2: Just activate WSL in the Windows Features (faster via the following commands) 
  - ``dism.exe /online /enable-feature /featurename:Microsoft-Windows-Subsystem-Linux /all /norestart``
  - ``dism.exe /online /enable-feature /featurename:VirtualMachinePlatform /all /norestart``
  - these are from the link from previous post.
- Step 3: Do not download anything fancy and just use this command and it should update to the latest WSL 2
  - ``wsl --update``
  - if system updates are not installed properly, this will get you 5.10.xx line, otherwise will properly get the latest version available (today it was 2.2.xx line).

PS: In the future, this discussion may go longer and this post might not be easy to find. Please warn user with the issue having stuck at 5.10.xx line to first update their system to latest or to the ones mentioned above. Only then discuss other issues, possibly tied to wsl.conf file, which probably we have solution in previous posts. This was long to read but I hope I was helpful.

**Comment by nicomollet** [comment]
I still had issues with `systemctl`.
I removed the `/etc/wsl.conf` file and recreated it again with nano inside WSL and it worked.

**Comment by hkn-tr** [comment]
Systemd is not native in WSL or WSL2, 
There may be a solution but I would not rather to hack.
https://github.com/shayne/wsl2-hacks

**Comment by yilmazdurmaz** [comment]
I am writing this post to ping you all for the moment for awareness, and I will later edit my previous post and add this as a warning and delete this one after a week or so to clear space.

I had another experience installing WSL 2.x store versions.  I installed Windows 22H2 on a USB drive through `dism /Apply-Image` method. (this makes it a "Windows-to-go")

After installing all the updates available through Windows Update, I tried to get WSL with the steps I previously followed in my above post. Unfortunately, this did not work as I expected.

There were only 3 system updates, ``KB4023057(2023-10)``, ``KB5033052(2024-01)``, and ``KB5001716(2024-05)``. Altough the system was up to date, using `wsl --update` command did not get the 5.15 version, and stayed in the old version line.

I had to download and install ``KB5020030``, and WSL updated smoothly after that.

**Comment by IsraelBoka** [comment]
It worked after an update `wsl --update`, i tried all the above without success

**Comment by Rollingbutter** [comment]
> [@MeechaPooch](https://github.com/MeechaPooch) you seem you have edited the config file to add default user. My guess is that you used an editor that changed the line endings from linux format (LF-0A) to dos format (CRLF-0D0A). system fails to load the config file in this case.
> 
> try `hexdump /etc/wsl.conf` and see if this is the case (does Ubuntu have hexdump by default? you may need to install it)
> 
> Fixing it depends on the editors to use and their settings for this format, and it will be long to give instructions here. Please search the internet about fixing this line ending problem.
It really works when I change CRLF-0D0A into LF-0A. Thank you so much, I will go crazy without your help.

**Comment by CABLE353** [comment]
I'm having this same issue too(being a newbie doesn't help either). Everything inside the /etc/wsl.conf file looks correct, but systemd doesn't start or "start with the system" even though it is set to do so in my wsl.conf file. acpid wasn't installed so I had to install it manually. I'm trying to run a full gnome desktop.

I'm running the Debian distro from the ms store, & on Win11 Pro with WSL2. I also tried to follow this tutorial right after Debian was set up:
[https://avivarma1.medium.com/setting-up-debian-on-wsl2-with-systemd-fb4831dd7b82](https://avivarma1.medium.com/setting-up-debian-on-wsl2-with-systemd-fb4831dd7b82)

**These are the contents of my wsl.conf file:**
```
# Set a command to run when a new WSL instance launches.
[boot]
sudo apt update && sudo apt upgrade
systemd=true

# Set whether WSL supports interop process like launching Windows apps and adding path variables. Setting these to false will block the launch of Windows processes and block adding $PATH environment variables.
[interop]
appendWindowsPath=true

# Automatically mount Windows drive when the distribution is launched
[automount]

# Set to true will automount fixed drives (C:/ or D:/) with DrvFs under the root directory set above. Set to false means drives won't be mounted automatically, but need to be mounted manually or with fstab.
enabled = false''
```

**Here's the hexdump for the file as well:**
```
0000000 2023 6553 2074 2061 6f63 6d6d 6e61 2064
0000010 6f74 7220 6e75 7720 6568 206e 2061 656e
0000020 2077 5357 204c 6e69 7473 6e61 6563 6c20
0000030 7561 636e 6568 2e73 0a0d 625b 6f6f 5d74
0000040 0a0d 7973 7473 6d65 3d64 7274 6575 0a0d
0000050 0a0d 2023 6553 2074 6877 7465 6568 2072
0000060 5357 204c 7573 7070 726f 7374 6920 746e
0000070 7265 706f 7020 6f72 6563 7373 6c20 6b69
0000080 2065 616c 6e75 6863 6e69 2067 6957 646e
0000090 776f 2073 7061 7370 6120 646e 6120 6464
00000a0 6e69 2067 6170 6874 7620 7261 6169 6c62
00000b0 7365 202e 6553 7474 6e69 2067 6874 7365
00000c0 2065 6f74 6620 6c61 6573 7720 6c69 206c
00000d0 6c62 636f 206b 6874 2065 616c 6e75 6863
00000e0 6f20 2066 6957 646e 776f 2073 7270 636f
00000f0 7365 6573 2073 6e61 2064 6c62 636f 206b
0000100 6461 6964 676e 2420 4150 4854 6520 766e
0000110 7269 6e6f 656d 746e 7620 7261 6169 6c62
0000120 7365 0d2e 5b0a 6e69 6574 6f72 5d70 0a0d
0000130 7061 6570 646e 6957 646e 776f 5073 7461
0000140 3d68 7274 6575 0a0d 0a0d 2023 7541 6f74
0000150 616d 6974 6163 6c6c 2079 6f6d 6e75 2074
0000160 6957 646e 776f 2073 7264 7669 2065 6877
0000170 6e65 7420 6568 6420 7369 7274 6269 7475
0000180 6f69 206e 7369 6c20 7561 636e 6568 0d64
0000190 5b0a 7561 6f74 6f6d 6e75 5d74 0a0d 0a0d
00001a0 2023 6553 2074 6f74 7420 7572 2065 6977
00001b0 6c6c 6120 7475 6d6f 756f 746e 6620 7869
00001c0 6465 6420 6972 6576 2073 4328 2f3a 6f20
00001d0 2072 3a44 292f 7720 7469 2068 7244 4676
00001e0 2073 6e75 6564 2072 6874 2065 6f72 746f
00001f0 6420 7269 6365 6f74 7972 7320 7465 6120
0000200 6f62 6576 202e 6553 2074 6f74 6620 6c61
0000210 6573 6d20 6165 736e 6420 6972 6576 2073
0000220 6f77 276e 2074 6562 6d20 756f 746e 6465
0000230 6120 7475 6d6f 7461 6369 6c61 796c 202c
0000240 7562 2074 656e 6465 7420 206f 6562 6d20
0000250 756f 746e 6465 6d20 6e61 6175 6c6c 2079
0000260 726f 7720 7469 2068 7366 6174 2e62 0a0d
0000270 6e65 6261 656c 2064 203d 6166 736c 0065
000027f
```

**Comment by yilmazdurmaz** [comment]
@CABLE353,  thanks for adding hex content. the last post before yours have quoted one of my answers about this issue: CRLF-0D0A

This thread is long so it is an easy-to-miss detail. It seems you had to use an editor outside WSL distro, or yours saved it in DOS CRLF format.

open it in another editor and make sure you save the file using only UNIX LF line endings. or check which editor options you need to use for you editor.

**Comment by CABLE353** [comment]
Well I did use mousepad inside the WSL environment. If I were to use mousepad again, how would I get the file to save in that "state"?

**Comment by yilmazdurmaz** [comment]
> Well I did use mousepad inside the WSL environment. If I were to use mousepad again, how would I get the file to save in that UNIX LF "state"?

@CABLE353, I haven't used it before, but from its source code, it seems to have this item/statement in one of its menus or settings:
 - Line Ending > Unix (_LF) 
 - with tooltip "Set the line ending of the document to Unix (LF)"

I haven't tried before, but I guess it would be nice to have one-liner codes with `sed` and/or `awk` to fix this CRLF fast from console. 
Anyone?

**Comment by bocki** [comment]
I was spending an hour trying to get systemd working on an old WSL install of Arch. /etc/wsl.conf was perfectly fine, yet systemd did not start. 

For anyone who is convinced that everything else is right: You need a symlink in `/sbin` (or, well, `/usr/bin`), which is part of new Arch installs but not my old one:
`ln -s ../lib/systemd/systemd init`

systemd works like a charm now. :)

**Comment by yomezza** [comment]
> your wsl version is 1.1.3.0 which means you already have a compatible version. that KB update possibly fails since you try to use arm64 on x64, or vice versa (you downloaded a wrong version).
> 
> > Tried with ubuntu 20.04 and upgraded to 22.04 but still didn't work
> 
> each distro is a new virtual machine, and each machine must be configured separately to start with systemd, then must be restarted.
> 
> ```
> wsl --list -v # need version 2
> wsl --set-version distronameyouset 2
> wsl -d distronameyouset
> sudo nano /etc/wsl.conf # or any editor that uses LF, or 0A, as new line
> exit
> wsl --terminate distronameyouset # shutdown single
> # wsl --shutdown # shutdown all
> wsl -d distronameyouset
> ```
> 
> there is one thing I am not sure about: this might require an x64 machine but you use x86 (32bit). and one more possible thing is, I don't know if has any relation, is WSL versus WSL2; can't say if 1.1.3.0 is possible with WSL but worth checking. try the above way first, then check these possibilities.
> 
> EDIT: After receiving the next answer, I put the "set-version" step in case someone in the future reads these steps. we need WSL2.

Thanks!
